src/Disks/IO/ThreadPoolRemoteFSReader.h

```diff
@@ -1,38 +1,39 @@
 #pragma once
 
 #include <IO/AsynchronousReader.h>
-#include <IO/SeekableReadBuffer.h>
 #include <Common/ThreadPool.h>
 #include <Disks/IO/ReadBufferFromRemoteFSGather.h>
 
-
 namespace DB
 {
 
+template <class Reader>
 class ThreadPoolRemoteFSReader : public IAsynchronousReader
 {
-
-private:
-    ThreadPool pool;
-
 public:
+    using ReadResult = IAsynchronousReader::Result;
+
     ThreadPoolRemoteFSReader(size_t pool_size, size_t queue_size_);
 
     std::future<Result> submit(Request request) override;
 
-    struct RemoteFSFileDescriptor;
+private:
+    ThreadPool pool;
 };
 
 
-struct ThreadPoolRemoteFSReader::RemoteFSFileDescriptor : public IFileDescriptor
+template <class Reader>
+class RemoteFSFileDescriptor : public IAsynchronousReader::IFileDescriptor
 {
 public:
-    explicit RemoteFSFileDescriptor(std::shared_ptr<ReadBufferFromRemoteFSGather> reader_) : reader(reader_) {}
+    using ReadResult = IAsynchronousReader::Result;
+
+    explicit RemoteFSFileDescriptor(std::shared_ptr<Reader> reader_) : reader(std::move(reader_)) { }
 
-    ReadBufferFromRemoteFSGather::ReadResult readInto(char * data, size_t size, size_t offset, size_t ignore = 0);
+    ReadResult readInto(char * data, size_t size, size_t offset, size_t ignore = 0);
 
 private:
-    std::shared_ptr<ReadBufferFromRemoteFSGather> reader;
+    std::shared_ptr<Reader> reader;
```

Comment:
May be we can make it std::shared_ptr<ReadBufferFromFileBase> instead of adding templates? I think there will be a few adjustments in buffers to allow that.

Reply:
I will try.